File: src/main/java/org/taktik/icure/services/external/http/websocket/KmehrFileOperation.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.http.websocket;

import java.io.IOException;
import java.io.Serializable;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Future;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import org.eclipse.jetty.websocket.api.Session;
import org.taktik.icure.services.external.api.AsyncDecrypt;
import org.taktik.icure.services.external.rest.v1.dto.embed.ServiceDto;

public class KmehrFileOperation extends BinaryOperation implements AsyncDecrypt {
	private Map<String,DecodingSession> decodingSessions = new HashMap<>();

	public KmehrFileOperation(WebSocket webSocket, Gson gsonMapper) {
		super(gsonMapper, webSocket);
	}

	@Override
	public <K extends Serializable> Future<List<K>> decrypt(List<K> encrypted, Class<K> clazz) throws IOException {
		Message message = new Message<>("decrypt", clazz.getSimpleName(), UUID.randomUUID().toString(), encrypted);

		CompletableFuture<List<K>> future = new CompletableFuture<>();
		DecodingSession<K> decodingSession = new DecodingSession<>(future, clazz);
		decodingSessions.put(message.getUuid(), decodingSession);

		webSocket.getRemote().sendString(gsonMapper.toJson(message));

		return future;
	}

	@Override
	public void handle(String message) {
		JsonParser parser = new JsonParser();
		JsonObject dto = parser.parse(message).getAsJsonObject();

		if (dto.get("command").getAsString().equals("decryptResponse")) {
			DecodingSession decodingSession = decodingSessions.get(dto.get("uuid").getAsString());
            if (decodingSession != null) {
                decodingSession.getFuture().complete(StreamSupport.stream(dto.get("body").getAsJsonArray().spliterator(), false).map(e -> {
                    try {
                        return gsonMapper.fromJson(e.getAsJsonObject(), decodingSession.getClazz());
                    } catch (com.google.gson.JsonSyntaxException ee) {
                        return null;
                    }
                }).collect(Collectors.toList()));
            }
		}
	}

	private class DecodingSession<K extends Serializable> {
		CompletableFuture<List<K>> future;
		Class<K> clazz;

		DecodingSession(CompletableFuture<List<K>> future, Class<K> clazz) {
			this.future = future;
			this.clazz = clazz;
		}

		public CompletableFuture<List<K>> getFuture() {
			return future;
		}

		public Class<K> getClazz() {
			return clazz;
		}
	}
}

File: src/main/kotlin/org/taktik/icure/services/external/http/websocket/KmehrFileOperation.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.services.external.http.websocket

import java.io.IOException
import java.io.Serializable
import java.time.Duration
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeoutException
import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.coroutines.reactive.awaitFirst
import kotlinx.coroutines.reactive.awaitFirstOrNull
import org.apache.commons.logging.LogFactory
import org.springframework.web.reactive.socket.WebSocketSession
import org.taktik.icure.be.ehealth.logic.kmehr.v20131001.KmehrExport
import org.taktik.icure.services.external.api.AsyncDecrypt
import reactor.core.publisher.Mono

class KmehrFileOperation(webSocket: WebSocketSession, objectMapper: ObjectMapper) : BinaryOperation(webSocket, objectMapper), AsyncDecrypt {
	private val log = LogFactory.getLog(KmehrExport::class.java)
	private val decodingSessions: MutableMap<String?, DecodingSession<*>> = HashMap()

	@Throws(IOException::class)
	override suspend fun <K : Serializable?> decrypt(encrypted: List<K>, clazz: Class<K>): List<K> {
		val message: Message<*> = Message("decrypt", clazz.simpleName, UUID.randomUUID().toString(), encrypted)
		val future = CompletableFuture<List<K>>()
		val decodingSession = DecodingSession(future, clazz)
		decodingSessions[message.uuid] = decodingSession
		val jsonMessage = objectMapper.writeValueAsString(message)
		val wsMessage = if (jsonMessage.length > 65000) webSocket.binaryMessage { it.wrap(jsonMessage.toByteArray(Charsets.UTF_8)) } else webSocket.textMessage(objectMapper.writeValueAsString(message))
		webSocket.send(Mono.just(wsMessage)).awaitFirstOrNull()
		return try {
			Mono.fromFuture(future).timeout(Duration.ofSeconds(120)).awaitFirst()
		} catch (toe: TimeoutException) {
			decodingSessions.remove(message.uuid)
			listOf()
		}
	}

	override fun <K : Serializable> handle(message: String?) {
		val dto = try {
			objectMapper.readTree(message)
		} catch (e: Exception) {
			log.error("Cannot parse because of $e. Object is: $message", e)
			throw (e)
		}
		if (dto["command"].asText() == "decryptResponse") {
			val uuid = dto["uuid"].asText()
			val decodingSession = decodingSessions[uuid] as DecodingSession<K>
			val result = dto["body"].mapNotNull { jsonObject ->
				try {
					val value = objectMapper.treeToValue<K>(jsonObject, decodingSession.clazz)
					value
				} catch (ee: Exception) {
					log.error("Cannot parse because of $ee")
					null
				}
			}
			decodingSession.future.complete(result)
			decodingSessions.remove(uuid)
		}
	}

	private inner class DecodingSession<K : Serializable?>(var future: CompletableFuture<List<K>>, var clazz: Class<K>)
}
